// @mapbox/search-js-react/web/Geocode.tsx
import React, { useRef, useEffect, useState } from 'react';
import mapboxgl from 'mapbox-gl';

import { Geocoder } from '../src/components/Geocoder';
import { MapboxGeocoder } from '@mapbox/search-js-web';
import { useGeocodingCore } from '../src';

const ACCESS_TOKEN = MAPBOX_ACCESS_TOKEN;

export function Geocode(): React.ReactElement {
  const [map, setMap] = useState(null);
  const [value, setValue] = useState('');

  const geocodeRef = useRef(null);

  const mapContainerRef = useRef(null);

  const geocodingCore = useGeocodingCore({ accessToken: ACCESS_TOKEN });

  const runGeocodingSuggest = React.useCallback(async () => {
    const response = await geocodingCore.suggest('1600 pennsylvania ave nw');
    console.log(response);
  }, [geocodingCore]);

  useEffect(() => {
    mapboxgl.accessToken = ACCESS_TOKEN;

    const map = new mapboxgl.Map({
      container: mapContainerRef.current,
      style: 'mapbox://styles/mapbox/streets-v11',
      center: [-73.943, 40.7789],
      zoom: 11
    });

    setMap(map);

    map.addControl(new mapboxgl.FullscreenControl());

    const mapSearchControl = new MapboxGeocoder();
    mapSearchControl.accessToken = ACCESS_TOKEN;
    mapSearchControl.mapboxgl = mapboxgl;
    mapSearchControl.marker = { color: 'orange' };
    map.addControl(mapSearchControl);

    // geocodeRef.current.search('central park, new york ny');

    // Clean up on unmount
    return () => map.remove();
  }, []);

  return (
    <div
      style={{
        height: '600px',
        width: '100%',
        margin: '30px',
        display: 'flex',
        flexDirection: 'column'
      }}
    >
      <div style={{ margin: '24px 12px' }}>
        <Geocoder
          accessToken={ACCESS_TOKEN}
          value={value}
          placeholder="Search for an address"
          onChange={(v) => {
            setValue(v);
            console.log('change', v);
          }}
          onSuggest={(res) => console.log('suggest', res)}
          onSuggestError={(err) => console.log('error', err)}
          onRetrieve={(res) => console.log('retrieve', res)}
          onClear={() => console.log('clear')}
          map={map}
          marker={true}
          mapboxgl={mapboxgl}
          ref={geocodeRef}
        />
      </div>
      <br></br>
      <div
        className="map-container"
        style={{ height: '100%', width: '100%' }}
        ref={mapContainerRef}
      />
      <button type="button" onClick={runGeocodingSuggest}>
        useGeocodingCore Test
      </button>
    </div>
  );
}
